csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -44,10 +44,10 @@ public InferenceSession(string modelPath, SessionOptions options)
             _nativeHandle = IntPtr.Zero;
             try
             {
-                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-                    NativeApiStatus.VerifySuccess(NativeMethods.OrtCreateSessionW(envHandle, modelPath, options.NativeHandle, out _nativeHandle));
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    NativeApiStatus.VerifySuccess(NativeMethods.OrtCreateSession(envHandle, System.Text.Encoding.Unicode.GetBytes(modelPath), options.NativeHandle, out _nativeHandle));
                 else
-                    NativeApiStatus.VerifySuccess(NativeMethods.OrtCreateSession(envHandle, modelPath, options.NativeHandle, out _nativeHandle));
+                    NativeApiStatus.VerifySuccess(NativeMethods.OrtCreateSession(envHandle, System.Text.Encoding.UTF8.GetBytes(modelPath), options.NativeHandle, out _nativeHandle));
```

Comment:
On Linux, we can't assume it's utf-8. The encoding is controlled by locale.

Reply:
This is mostly for byte[] conversion purposes -- the C-API uses char *...so we need to convert as if it's UTF-8 into byte representation, and let the onnxruntime interpret the locale. The exact final encoding will be interpreted by Onnxruntime and environment locale, as long as it's sent in as a char*, and not a wchar_t*.

Reply:
This assume works well for C#, but how about the C/C++ users ? It will break their code. Like the fns_candy demo in this repo, I have to do extra coding conversion before calling onnxruntime APIs, it sounds inconvenient.

Reply:
The code change is in C#. How would C/C++ users be impacted?

C# stores strings in Unicode-UTF16 (2 bytes). The second byte is almost always null-terminated, which means the model path will always look like it's 1 character long. The needs to be converted into an 8-bit encoding so Linux can interpret as char*.

The available C# encodings are shown below.
https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding?view=netframework-4.7.2

The 7-bit, 16-bit and 32-bit encodings don't work in Linux. Only ASCII and UTF-8 seem work, but ASCII will almost certainly have limitations (so should not use).

Reply:
The C API(e.g. OrtCreateSession) can not assume its input string is in utf-8. The actual encoding is controlled by locale.

Reply:
Will set up a discussion for  localization/encoding, vs LPSTR vs LPWSTR for C#/C++ interop. May require a followup PR.